### DISCUSSION TITLE
Heterogeneous dataloader with batches

### DISCUSSION THREAD

**Discussion by LaurentBerder** [pyg-team/pytorch_geometric#6350] [Q&A]
Hi,

I am spending a lot of time using pytorch_geometric lately, and want to take the occasion of my post here to thank you guys for all the work you've done in the development.

I am modeling a phenomenon with 5 types of nodes and 7 types of edges. I've started by creating a full graph, but the model training crashes, so I'm looking into batches.
I first started looking into `HGTLoader` and `NeighboorLoader`, but they picked nodes a bit too much at random. Since my data is rather organic (the phenomenon will basically always have the same sort of shape similar to the below), I am now creating a "mini-graph" for each phenomenon.
![mini_graph](https://i.stack.imgur.com/cL3tQ.png)

Once I created all the mini_graphs, I group them in a DataLoader `train_loader = DataLoader(train_data, batch_size=100, shuffle=True)`, ensuring that all of my phenomenons (with all their node types) would be complete in each batch.

The problem I have now in launching the training loop is that the batches don't contain all the nodes of the graph, and therefore I get an error on the node_indexes being too large:
```python
ValueError: Found indices in 'edge_index' that are larger than 99 (got 7430). Please ensure that all indices in 'edge_index' point to valid indices in the interval [0, 100) in your node feature matrix and try again.
```

Indeed, since my whole graph contains 9K nodes of type 1, i might have, in edge_index, numbers higher than 99 even though the graph presented in the batch only contains 100 nodes of type 1.

How do you think I could get out of this predicament?

**Comment by rusty1s** [comment]
How do you create your sub-graphs? Keep in mind that the edge indices of each subgraph needs to be mapped to new indices (from `0` to `num_subgraph_nodes -1 `.

**Reply by LaurentBerder** [reply]
I create the subgraphs as a list of `HeteroData`.

> Keep in mind that the edge indices of each subgraph needs to be mapped to new indices (from 0 to num_subgraph_nodes -1

That's what I was afraid of. How will the model learn that a node is encountered in other subgraphs if it bears different indices each time?
I was rather hoping that in a `DataLoader`, it would be possible to have any kind of index, as long as on the whole data, the `from 0 to num_subgraph_nodes -1` rule was respected.

**Reply by rusty1s** [reply]
Each subgraph `data` object needs to be well-defined (`data.subgraph` should take care of this already), the indices can be different since GNNs are permutation-invariant by design.

**Comment by LaurentBerder** [comment]
I understand.

But I have a follow-up question then: once I trained the model and made predictions on the re-indexed nodes, how do I get back the original indices to know which node has been predicted?

For example, one of the node types is booking, and another is client. I stripped both of their unique IDs in the node.x features, and can only count on their index to identify them. So how can I then see what the forecasts are for all the bookings of a certain client (client which very likely may appear in different subgraphs).

**Reply by rusty1s** [reply]
You can maintain the mapping in a data attribute, e.g.:
```python
data["client"].node_id = torch.arange(data["client"].num_nodes)
data.subgraph(...)

**Reply by LaurentBerder** [reply]
Awesome, that's kinda what I thought about just after I wrote the question.

Thanks a lot for your help (and amazing reactivity)!

**Reply by dinhngoc267** [reply]
Thanks @LaurentBerder for raising this question. and thanks @rusty1s for your answer. 

But is it because we compact the nodes and edges as dense matrix? Can we represent it as sparse matrix where the index of the node is consistent between subgraphs. As a case just appear in my mind that many subgraph can have many same nodes and for example we do not have these nodes's features and we initialize an nn.Embedding to embed the nodes and train an end-to-end manner with the pipeline graph learning. So can a sparse matrix be a solution for this? @rusty1s
